packages/core/src/services/chains/ChainStreamConsumer/index.ts

```diff
@@ -0,0 +1,173 @@
+import { MessageRole } from '@latitude-data/compiler'
+
+import {
+  ChainEvent,
+  ChainEventTypes,
+  ChainStepResponse,
+  RunErrorCodes,
+  StreamEventTypes,
+  StreamType,
+} from '../../../constants'
+import { objectToString } from '../../../helpers'
+import { Config } from '../../ai'
+import { ChainError } from '../ChainErrors'
+import { ValidatedStep } from '../ChainValidator'
+
+export function parseResponseText(response: ChainStepResponse<StreamType>) {
+  if (response.streamType === 'object') return response.text || ''
+
+  const text = response.text
+  const hasTextResponse = text && text.length > 0
+  if (hasTextResponse) return text
+
+  if (!response.toolCalls.length) return text ?? ''
+
+  const toolsString = objectToString(
+    response.toolCalls,
+    'Error parsing tool calls',
```

Comment:
is this a fallback error? shouldn't we throw if we can't parse toolcalls? it should be a valid json

Reply:
This method was already doing a JSON.stringify like you. The difference it catch the error if fails. 
This was not catching the fail of stringify. I just used our existing `objectToString` and pass a custom fallback message
https://github.com/latitude-dev/latitude-llm/pull/360/files#diff-401ba2f40b8794f8afeaef9441869e1c2c3ad9fad535909f156c47d60c7ba061R277